### PR TITLE
feat: add support for incremental builds

### PR DESCRIPTION
**Pull request by alan-agius4**
## I'm submitting a...

```
[ ] Bug Fix
[x] Feature
[ ] Other (Refactoring, Added tests, Documentation, ...)
```

## Checklist

- [x] Commit Messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
- [x] Tests for the changes have been added

## Description

This PR, bring incremental builds, in `CLI` you can enable the watch, by providing the `-w` or `--watch` arguments. While for the `API`, you can use the `withOptions` method example.

CLI
```
ng-packagr --project ng-packagr.js --watch
```

API
```
ngPackage
  .ngPackagr()
  .forProject('ng-package.js')
  .withOptions({watch: true})
  .build()
```

This requires `enableResourceInlining` to be enabled and also this is a sort of MVP. There are more optimization that need to be done.

Such as;
- Cache Compiler Hosts for later usage
- Cache Program for later usage
- Build only the entry pointthat changed
- Provide a callback on every re-compilation
- Clean respective `.d.ts` from when `.ts` changes

Closes: #828, Closes: #743 and Closes: #635,
Closes: https://github.com/angular/angular-cli/issues/10643

Depends on https://github.com/dherges/ng-packagr/pull/875 and https://github.com/dherges/ng-packagr/pull/872

This is quite experimental and would love to get some feedback from users

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: renovate.json [hunk 1]
```diff
@@ -7,28 +7,25 @@
...
-  "packageFiles": [
-    "package.json",
-    "integration/consumers/ng-cli/package.json"
-  ],
+  "packageFiles": ["package.json", "integration/consumers/ng-cli/package.json"],
   "packageRules": [
     {
-      "packagePatterns": ["^@angular\/.*"],
+      "packagePatterns": ["^@angular/.*"],
       "groupName": "angular",
```
Comment (alan-agius4) [inline renovate.json]: Damn hook, need to remember this and add it in config

**Comment by alan-agius4** [comment]
Ill resolve the conflicts later, as most of stuff conflicting has been removed with this PR.

File: src/lib/ng-v5/entry-point/resources/stylesheet-processor.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import * as stylus from 'stylus';
+
+export class StylesheetProcessor {
+  private postCssProcessor: postcss.Processor;
+
+  constructor(readonly basePath: string, readonly cssUrl?: CssUrl, readonly styleIncludePaths?: string[]) {
+    this.postCssProcessor = this.createPostCssProcessor(basePath, cssUrl, styleIncludePaths);
+  }
+
+  process(filePath: string, content: string) {
```
Comment (dherges) [inline src/lib/ng-v5/entry-point/resources/stylesheet-processor.ts]: Note: we made stylesheet processing sync, because it's triggered through the TypeScript transformations (which is a sync API).

File: src/lib/ts/cache-compiler-host.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    },
+
+    readResource: (fileName: string) => {
+      const cache = sourcesFileCache.getOrCreate(fileName);
+      if (cache.content === undefined) {
+        // todo: transform styles here.
+        // the empty string is needed because of include paths file's won't be resolved properly.
+        cache.content = compilerHost.readFile.call(this, fileName);
+        if (!/(html|htm)$/.test(path.extname(fileName))) {
+          cache.content = stylesheetProcessor.process(fileName, cache.content);
```
Comment (dherges) [inline src/lib/ts/cache-compiler-host.ts]: Note: replaces the `stylesheet.transform` from before. With that change, stylesheet processing needs to by synchronous.

**Comment by alan-agius4** [comment]
@dherges bump

**Comment by alan-agius4** [comment]
@dherges you think you can find some time to review this and so we get this merged?

**Comment by dherges** [comment]
I try to. The timing about time is that it only goes forward and I'm not having much for the GitHub things at the moment.

Still this is a big change and I will not understand every detail in the code. I'll just be able to try it out from a user perspective and then merge it in - or not. It's either a very very huge step or get the packagr into a dead end.

It's not got nothing to do with your code. It's about the custom pipeline and build graph and stuff. In the beginning I felt like a custom layer for tool orchestrating was neccessary - because of ngc, rollup, scss, less and the tools were changing a lot. I recently figured out that rollup is capable of writing multiple output bundles from the same input. With the `enableResourceInling` I think it's now also capable to do a full Angular build (with packaging) with only rollup and plugins.

Long story short: on the long run, a lot of the cutom code in ng-packagr is going to go away. But well, adding the incremental builds and watch mode is a nice thing!

File: src/lib/ng-v5/package.transform.ts [hunk 4]
```diff
@@ -73,6 +80,67 @@ export const packageTransformFactory = (
...
+  entryPointTransform: Transform
+) => (source$: Observable<BuildGraph>): Observable<BuildGraph> => {
+  const CompleteWaitingForFileChange = '\nCompilation complete. Watching for file changes...';
+  const FileChangeDetected = '\nFile change detected. Starting incremental compilation...';
+  const FailedWaitingForFileChange = '\nCompilation failed. Watching for file changes...';
+
+  return source$.pipe(
+    switchMap(graph => {
+      const { data } = graph.find(isPackage) as PackageNode;
+      return createFileWatch(data.src, [data.dest]).pipe(
```
Comment (dherges) [inline src/lib/ng-v5/package.transform.ts]: Yes, nice. TO summarize what the observable pipeline looks:

```ts
of(new BuildGraph()).pipe(
  switchMap((..) => createFileWatch()) //  file watch emits on file changes
     .pipe(
        tap(() => {}) // mark nodes as dirty
      ),
   switchMap(() => doTheBuild()) // run the compilation and bundling and packaging pipeline
)
```

File: src/lib/commands/build.command.ts [hunk 5]
```diff
@@ -19,4 +21,5 @@ export interface CliArguments {
 export const build: Command<CliArguments, void> = opts =>
   ngPackagr()
     .forProject(opts.project)
+    .withOptions({ watch: opts.watch })
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply src/lib/commands/build.command.ts]: I though about this for long and I was not sure which one was the best.
